### DISCUSSION TITLE
Sail - MinIO - Access to Files in Bucket From Browser

### DISCUSSION THREAD

**Discussion by falconsmilie** [laravel/framework#40188] [Q&A]
Hi.

My aim is to use [Sail](https://laravel.com/docs/8.x/sail) with MinIO for local dev.

In an attempt to troubleshoot in a clean install; I have a default Laravel install (`"laravel/framework": "^8.75"`) with Sail (`"laravel/sail": "^1.0.1"`).

One odd thing is I had to require `"league/flysystem-aws-s3-v3": "~1.0"`, newer versions did not install.

In any case, I am able to write an upload feature that saves the files to the MinIO bucket storage;

```
// Controller
$path = $request->file('file')->store('content');
```

<img width="640" alt="Screen Shot 2021-12-28 at 11 37 24" src="https://user-images.githubusercontent.com/1234911/147557976-30fa1910-103f-4109-b2a6-b9405d2a43b7.png">

The issue I am facing is I can't view the files in the browser. The `$path`above is correct, but when added to `<img>` the display is the broken image, and when attempting to navigate directly to the image;

```http://minio:9000/local/content/X8l8M5TyLwDQsijHq9bojrTvLfh9hNfT9C9BP4GF.jpg```

 the browser always says;

```This site can’t be reached minio’s server IP address could not be found.```

If i switch out the domain to `localhost` or `127.0.0.1` i get access denied XML response.

I am developing on a MacBook Air M1.

Can someone suggest where I have gone wrong?

**Comment by falconsmilie** [comment]
The only way i can get it to work is by adding this to the `~/private/etc/hosts`;

`127.0.0.1 minio`

**Reply by hlorofos** [reply]
That drives me nuts. Thanks for helping by posting your comment as the solution!

**Reply by ekpono** [reply]
How to I do about adding theses?

**Reply by falconsmilie** [reply]
Add `127.0.0.1 minio` to your `hosts` file. Google can help with where the `hosts` file is located on whatever operating system you are using.

**Reply by trane77** [reply]
Beautiful, thanks

**Comment by kassah** [comment]
With Laravel 11 at least, you can set the filesystem config S3 option of [http_handler](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_configuration.html#http-handler) and pass in the CONNECT_TO param to the underlying curl, redirecting the request at the last moment to minio within PHP, which leaves things like signed urls returning the endpoint as actually localhost, which is where the browser can reach it.

For example:
```php
's3' => [
    'driver' => 's3',
    'key' => env('AWS_ACCESS_KEY_ID'),
    'secret' => env('AWS_SECRET_ACCESS_KEY'),
    'region' => env('AWS_DEFAULT_REGION'),
    'bucket' => env('AWS_BUCKET'),
    'url' => env('AWS_URL'),
    'endpoint' => 'http://localhost:9000',
    'use_path_style_endpoint' => env('AWS_USE_PATH_STYLE_ENDPOINT', false),
    'throw' => false,
    'http_handler' => new \Aws\Handler\GuzzleV6\GuzzleHandler(
        new \GuzzleHttp\Client([
            'curl' => [
                CURLOPT_CONNECT_TO => ['localhost:9000:minio:9000'],
            ]
        ])
    ),
],
```

This requires the AWS client of V3 (for the [http_handler](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_configuration.html#http-handler) config param), PHP 7.0.7+ and curl 7.49.0+ and won't work with `artisan config:cache` as Laravel can't serialize the GuzzleHandler to a cached config.
